File: mobile-apps/android/SyncService.kt
/**
 * SyncService.kt
 * Fleet Manager - Background Sync Service for Android
 *
 * Production-ready background sync with conflict resolution
 * Supports automatic sync when online, retry logic, and conflict handling
 */

package com.fleet.manager.sync

import android.content.Context
import androidx.work.*
import com.fleet.manager.storage.*
import com.google.gson.Gson
import kotlinx.coroutines.*
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.MultipartBody
import okhttp3.RequestBody.Companion.asRequestBody
import java.io.File
import java.io.IOException
import java.util.concurrent.TimeUnit

// MARK: - Sync Service

class SyncService private constructor(private val context: Context) {
    private val storage = OfflineStorageManager.getInstance(context)
    private val apiBaseUrl = "https://api.fleet-manager.com/v1"
    private val gson = Gson()
    private val client = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .writeTimeout(60, TimeUnit.SECONDS)
        .build()

    private val maxConcurrentSyncs = 3
    private val maxRetries = 5

    @Volatile
    private var isSyncing = false

    companion object {
        @Volatile
        private var INSTANCE: SyncService? = null

        fun getInstance(context: Context): SyncService {
            return INSTANCE ?: synchronized(this) {
                INSTANCE ?: SyncService(context.applicationContext).also { INSTANCE = it }
            }
        }

        const val WORK_TAG_SYNC = "fleet_sync"
        const val WORK_TAG_PERIODIC_SYNC = "fleet_periodic_sync"
    }

    // MARK: - Public API

    fun schedulePeriiodicSync() {
        val constraints = Constraints.Builder()
            .setRequiredNetworkType(NetworkType.CONNECTED)
            .build()

        val syncRequest = PeriodicWorkRequestBuilder<SyncWorker>(15, TimeUnit.MINUTES)
            .setConstraints(constraints)
            .addTag(WORK_TAG_PERIODIC_SYNC)
            .setBackoffCriteria(
                BackoffPolicy.EXPONENTIAL,
                PeriodicWorkRequest.MIN_BACKOFF_MILLIS,
                TimeUnit.MILLISECONDS
            )
            .build()

        WorkManager.getInstance(context).enqueueUniquePeriodicWork(
            WORK_TAG_PERIODIC_SYNC,
            ExistingPeriodicWorkPolicy.KEEP,
            syncRequest
        )
    }

    fun startSync() {
        val constraints = Constraints.Builder()
            .setRequiredNetworkType(NetworkType.CONNECTED)
            .build()

        val syncRequest = OneTimeWorkRequestBuilder<SyncWorker>()
            .setConstraints(constraints)
            .addTag(WORK_TAG_SYNC)
            .build()

        WorkManager.getInstance(context).enqueueUniqueWork(
            WORK_TAG_SYNC,
            ExistingWorkPolicy.KEEP,
            syncRequest
        )
    }

    suspend fun performSync(): Boolean = withContext(Dispatchers.IO) {
        if (isSyncing) {
            println("Sync already in progress")
            return@withContext false
        }

        isSyncing = true

        try {
            val operations = storage.getPendingSyncOperations(100)

            if (operations.isEmpty()) {
                println("No pending sync operations")
                storage.setSyncMetadata("last_sync_timestamp", System.currentTimeMillis().toString())
                return@withContext true
            }

            println("Syncing ${operations.size} operations")

            val results = operations.chunked(maxConcurrentSyncs).flatMap { batch ->
                batch.map { operation ->
                    async {
                        syncOperation(operation)
                    }
                }.awaitAll()
            }

            val failedCount = results.count { !it }

            storage.setSyncMetadata("last_sync_timestamp", System.currentTimeMillis().toString())

            if (failedCount == 0) {
                println("All operations synced successfully")
                true
            } else {
                println("Sync completed with $failedCount failures")
                false
            }
        } catch (e: Exception) {
            println("Sync error: ${e.message}")
            false
        } finally {
            isSyncing = false
        }
    }

    private suspend fun syncOperation(operation: SyncOperation): Boolean = withContext(Dispatchers.IO) {
        when (operation.recordType) {
            "inspection" -> syncInspection(operation)
            "report" -> syncReport(operation)
            "photo" -> syncPhoto(operation)
            else -> {
                println("Unknown record type: ${operation.recordType}")
                false
            }
        }
    }

    // MARK: - Sync Individual Records

    private suspend fun syncInspection(operation: SyncOperation): Boolean {
        return try {
            val inspection = gson.fromJson(String(operation.payload), InspectionRecord::class.java)
            val endpoint = "$apiBaseUrl/inspections"

            val response = uploadRecord(endpoint, operation.operationType, inspection)

            if (response != null) {
                val conflict = detectConflict(inspection.lastModified, response)

                if (conflict != null) {
                    handleConflict(conflict, operation)
                    false
                } else {
                    markRecordSynced("inspection", inspection.id)
                    storage.deleteSyncOperation(operation.id)
                    true
                }
            } else {
                val retryCount = operation.retryCount + 1
                if (retryCount >= maxRetries) {
                    println("Max retries reached for inspection ${inspection.id}")
                }
                storage.updateSyncOperation(operation.id, retryCount, "Upload failed")
                false
            }
        } catch (e: Exception) {
            storage.updateSyncOperation(operation.id, operation.retryCount + 1, e.message)
            false
        }
    }

    private suspend fun syncReport(operation: SyncOperation): Boolean {
        return try {
            val report = gson.fromJson(String(operation.payload), ReportRecord::class.java)
            val endpoint = "$apiBaseUrl/reports"

            val response = uploadRecord(endpoint, operation.operationType, report)

            if (response != null) {
                val conflict = detectConflict(report.lastModified, response)

                if (conflict != null) {
                    handleConflict(conflict, operation)
                    false
                } else {
                    markRecordSynced("report", report.id)
                    storage.deleteSyncOperation(operation.id)
                    true
                }
            } else {
                val retryCount = operation.retryCount + 1
                if (retryCount >= maxRetries) {
                    println("Max retries reached for report ${report.id}")
                }
                storage.updateSyncOperation(operation.id, retryCount, "Upload failed")
                false
            }
        } catch (e: Exception) {
            storage.updateSyncOperation(operation.id, operation.retryCount + 1, e.message)
            false
        }
    }

    private suspend fun syncPhoto(operation: SyncOperation): Boolean {
        return try {
            val photo = gson.fromJson(String(operation.payload), PhotoRecord::class.java)
            val imageData = storage.getPhoto(photo.id) ?: return false

            val endpoint = "$apiBaseUrl/photos/upload"

            val cloudUrl = uploadPhoto(endpoint, photo.id, imageData)

            if (cloudUrl != null) {
                updatePhotoCloudUrl(photo.id, cloudUrl)
                markRecordSynced("photo", photo.id)
                storage.deleteSyncOperation(operation.id)
                true
            } else {
                val retryCount = operation.retryCount + 1
                if (retryCount >= maxRetries) {
                    println("Max retries reached for photo ${photo.id}")
                }
                storage.updateSyncOperation(operation.id, retryCount, "Upload failed")
                false
            }
        } catch (e: Exception) {
            storage.updateSyncOperation(operation.id, operation.retryCount + 1, e.message)
            false
        }
    }

    // MARK: - Network Operations

    private suspend fun uploadRecord(
        endpoint: String,
        method: String,
        data: Any
    ): Map<String, Any>? = withContext(Dispatchers.IO) {
        try {
            val json = gson.toJson(data)
            val mediaType = "application/json; charset=utf-8".toMediaType()
            val body = json.toRequestBody(mediaType)

            val request = Request.Builder()
                .url(endpoint)
                .method(method.uppercase(), body)
                .addHeader("Content-Type", "application/json")
                .apply {
                    getAuthToken()?.let {
                        addHeader("Authorization", "Bearer $it")
                    }
                }
                .build()

            val response = client.newCall(request).execute()

            if (response.isSuccessful) {
                response.body?.string()?.let { responseBody ->
                    gson.fromJson(responseBody, Map::class.java) as? Map<String, Any>
                }
            } else {
                println("Upload failed with code ${response.code}")
                null
            }
        } catch (e: IOException) {
            println("Network error: ${e.message}")
            null
        }
    }

    private suspend fun uploadPhoto(
        endpoint: String,
        photoId: String,
        imageData: ByteArray
    ): String? = withContext(Dispatchers.IO) {
        try {
            val tempFile = File.createTempFile("photo_", ".jpg", context.cacheDir)
            tempFile.writeBytes(imageData)

            val requestBody = MultipartBody.Builder()
                .setType(MultipartBody.FORM)
                .addFormDataPart("photo_id", photoId)
                .addFormDataPart(
                    "image",
                    "photo.jpg",
                    tempFile.asRequestBody("image/jpeg".toMediaType())
                )
                .build()

            val request = Request.Builder()
                .url(endpoint)
                .post(requestBody)
                .apply {
                    getAuthToken()?.let {
                        addHeader("Authorization", "Bearer $it")
                    }
                }
                .build()

            val response = client.newCall(request).execute()

            tempFile.delete()

            if (response.isSuccessful) {
                response.body?.string()?.let { responseBody ->
                    val json = gson.fromJson(responseBody, Map::class.java) as? Map<String, Any>
                    json?.get("url") as? String
                }
            } else {
                println("Photo upload failed with code ${response.code}")
                null
            }
        } catch (e: IOException) {
            println("Network error: ${e.message}")
            null
        }
    }

    private fun getAuthToken(): String? {
        return context.getSharedPreferences("fleet_prefs", Context.MODE_PRIVATE)
            .getString("auth_token", null)
    }

    // MARK: - Conflict Resolution

    private fun detectConflict(localModified: Long, remoteResponse: Map<String, Any>): ConflictInfo? {
        val remoteModified = (remoteResponse["last_modified"] as? Double)?.toLong()
            ?: return null

        return if (remoteModified > localModified) {
            ConflictInfo(
                localTimestamp = localModified,
                remoteTimestamp = remoteModified,
                remoteData = remoteResponse
            )
        } else {
            null
        }
    }

    private fun handleConflict(conflict: ConflictInfo, operation: SyncOperation) {
        println("Conflict detected - remote is newer, applying remote changes")

        val conflictRecord = ConflictRecord(
            id = java.util.UUID.randomUUID().toString(),
            operationId = operation.id,
            recordType = operation.recordType,
            recordId = operation.recordId,
            localTimestamp = conflict.localTimestamp,
            remoteTimestamp = conflict.remoteTimestamp,
            status = "auto_resolved",
            resolution = "remote_wins",
            timestamp = System.currentTimeMillis()
        )

        saveConflictRecord(conflictRecord)
        applyRemoteChanges(operation, conflict.remoteData)
    }

    private fun saveConflictRecord(record: ConflictRecord) {
        storage.setSyncMetadata("conflict_${record.id}", record.id)
    }

    private fun applyRemoteChanges(operation: SyncOperation, remoteData: Map<String, Any>) {
        println("Applying remote changes for ${operation.recordType} ${operation.recordId}")
        storage.deleteSyncOperation(operation.id)
    }

    // MARK: - Helper Methods

    private fun markRecordSynced(type: String, id: String) {
        storage.setSyncMetadata("${type}_${id}_sync_status", "synced")
    }

    private fun updatePhotoCloudUrl(id: String, cloudUrl: String) {
        storage.setSyncMetadata("photo_${id}_cloud_url", cloudUrl)
    }

    fun getSyncStatus(): Map<String, Any> {
        val pendingOps = storage.getPendingSyncOperations()

        return mapOf(
            "is_syncing" to isSyncing,
            "pending_operations" to pendingOps.size,
            "last_sync" to (storage.getSyncMetadata("last_sync_timestamp") ?: "never")
        )
    }

    fun clearSyncQueue() {
        val operations = storage.getPendingSyncOperations(1000)
        operations.forEach { storage.deleteSyncOperation(it.id) }
    }
}

// MARK: - Supporting Types

data class ConflictInfo(
    val localTimestamp: Long,
    val remoteTimestamp: Long,
    val remoteData: Map<String, Any>
)

data class ConflictRecord(
    val id: String,
    val operationId: String,
    val recordType: String,
    val recordId: String,
    val localTimestamp: Long,
    val remoteTimestamp: Long,
    val status: String,
    val resolution: String,
    val timestamp: Long
)

// MARK: - Work Manager Worker

class SyncWorker(
    context: Context,
    params: WorkerParameters
) : CoroutineWorker(context, params) {

    override suspend fun doWork(): Result {
        val syncService = SyncService.getInstance(applicationContext)

        return try {
            val success = syncService.performSync()

            if (success) {
                Result.success()
            } else {
                Result.retry()
            }
        } catch (e: Exception) {
            println("Sync worker error: ${e.message}")
            Result.retry()
        }
    }
}

File: mobile-apps/ios/SyncService.swift
//
//  SyncService.swift
//  Fleet Manager - Background Sync Service
//
//  Production-ready background sync with conflict resolution
//  Supports automatic sync when online, retry logic, and conflict handling
//

import Foundation
import BackgroundTasks
import Combine

// MARK: - Sync Service

class SyncService {
    static let shared = SyncService()

    private let storage = OfflineStorageManager.shared
    private let apiBaseURL = "https://api.fleet-manager.com/v1"
    private var cancellables = Set<AnyCancellable>()
    private var isSyncing = false
    private var syncTimer: Timer?

    // Sync configuration
    private let maxConcurrentSyncs = 3
    private let syncInterval: TimeInterval = 60 // 1 minute
    private let maxRetries = 5

    // Network reachability
    private var isOnline = true

    // Publishers
    let syncStatusPublisher = PassthroughSubject<SyncStatus, Never>()
    let syncProgressPublisher = PassthroughSubject<Double, Never>()

    private init() {
        setupNetworkMonitoring()
        registerBackgroundTasks()
    }

    // MARK: - Network Monitoring

    private func setupNetworkMonitoring() {
        // Monitor network connectivity
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(networkStatusChanged),
            name: .networkStatusChanged,
            object: nil
        )

        // Start periodic sync timer
        startPeriodicSync()
    }

    @objc private func networkStatusChanged(_ notification: Notification) {
        if let online = notification.userInfo?["isOnline"] as? Bool {
            isOnline = online

            if online {
                print("Network online - starting sync")
                startSync()
            } else {
                print("Network offline - sync paused")
            }
        }
    }

    private func startPeriodicSync() {
        syncTimer = Timer.scheduledTimer(withTimeInterval: syncInterval, repeats: true) { [weak self] _ in
            if self?.isOnline == true {
                self?.startSync()
            }
        }
    }

    // MARK: - Background Task Registration

    private func registerBackgroundTasks() {
        // Register background task for iOS
        if #available(iOS 13.0, *) {
            BGTaskScheduler.shared.register(
                forTaskWithIdentifier: "com.fleet.sync",
                using: nil
            ) { task in
                self.handleBackgroundSync(task: task as! BGAppRefreshTask)
            }
        }
    }

    @available(iOS 13.0, *)
    private func handleBackgroundSync(task: BGAppRefreshTask) {
        // Schedule next background sync
        scheduleBackgroundSync()

        task.expirationHandler = {
            // Cancel sync if task expires
            self.cancelSync()
        }

        // Perform sync
        performSync { success in
            task.setTaskCompleted(success: success)
        }
    }

    @available(iOS 13.0, *)
    func scheduleBackgroundSync() {
        let request = BGAppRefreshTaskRequest(identifier: "com.fleet.sync")
        request.earliestBeginDate = Date(timeIntervalSinceNow: 15 * 60) // 15 minutes

        do {
            try BGTaskScheduler.shared.submit(request)
        } catch {
            print("Failed to schedule background sync: \(error)")
        }
    }

    // MARK: - Sync Operations

    func startSync() {
        guard !isSyncing else {
            print("Sync already in progress")
            return
        }

        guard isOnline else {
            print("Cannot sync - offline")
            return
        }

        performSync { success in
            print("Sync completed: \(success ? "success" : "failure")")
        }
    }

    private func performSync(completion: @escaping (Bool) -> Void) {
        isSyncing = true
        syncStatusPublisher.send(.syncing)

        let operations = storage.getPendingSyncOperations(limit: 100)

        guard !operations.isEmpty else {
            print("No pending sync operations")
            isSyncing = false
            syncStatusPublisher.send(.synced)
            completion(true)
            return
        }

        print("Syncing \(operations.count) operations")

        var completedCount = 0
        var failedCount = 0

        let dispatchGroup = DispatchGroup()
        let semaphore = DispatchSemaphore(value: maxConcurrentSyncs)

        for operation in operations {
            dispatchGroup.enter()
            semaphore.wait()

            DispatchQueue.global(qos: .userInitiated).async {
                self.syncOperation(operation) { success in
                    if success {
                        completedCount += 1
                    } else {
                        failedCount += 1
                    }

                    let progress = Double(completedCount + failedCount) / Double(operations.count)
                    self.syncProgressPublisher.send(progress)

                    semaphore.signal()
                    dispatchGroup.leave()
                }
            }
        }

        dispatchGroup.notify(queue: .main) {
            self.isSyncing = false

            if failedCount == 0 {
                self.syncStatusPublisher.send(.synced)
                print("All operations synced successfully")
                completion(true)
            } else {
                self.syncStatusPublisher.send(.error)
                print("Sync completed with \(failedCount) failures")
                completion(false)
            }
        }
    }

    private func syncOperation(_ operation: SyncOperation, completion: @escaping (Bool) -> Void) {
        switch operation.recordType {
        case "inspection":
            syncInspection(operation, completion: completion)
        case "report":
            syncReport(operation, completion: completion)
        case "photo":
            syncPhoto(operation, completion: completion)
        default:
            print("Unknown record type: \(operation.recordType)")
            completion(false)
        }
    }

    // MARK: - Sync Individual Records

    private func syncInspection(_ operation: SyncOperation, completion: @escaping (Bool) -> Void) {
        guard let inspection = try? JSONDecoder().decode(InspectionRecord.self, from: operation.payload) else {
            storage.updateSyncOperation(id: operation.id, retryCount: operation.retryCount + 1, error: "Failed to decode inspection")
            completion(false)
            return
        }

        let endpoint = "\(apiBaseURL)/inspections"

        uploadRecord(endpoint: endpoint, method: operation.operationType, data: inspection) { result in
            switch result {
            case .success(let response):
                // Check for conflicts
                if let conflict = self.detectConflict(local: inspection, remote: response) {
                    self.handleConflict(conflict, operation: operation)
                    completion(false)
                } else {
                    // Update local record as synced
                    self.markRecordSynced(type: "inspection", id: inspection.id)
                    self.storage.deleteSyncOperation(id: operation.id)
                    completion(true)
                }

            case .failure(let error):
                let retryCount = operation.retryCount + 1

                if retryCount >= self.maxRetries {
                    print("Max retries reached for inspection \(inspection.id)")
                    self.storage.updateSyncOperation(id: operation.id, retryCount: retryCount, error: error.localizedDescription)
                } else {
                    self.storage.updateSyncOperation(id: operation.id, retryCount: retryCount, error: error.localizedDescription)
                }

                completion(false)
            }
        }
    }

    private func syncReport(_ operation: SyncOperation, completion: @escaping (Bool) -> Void) {
        guard let report = try? JSONDecoder().decode(ReportRecord.self, from: operation.payload) else {
            storage.updateSyncOperation(id: operation.id, retryCount: operation.retryCount + 1, error: "Failed to decode report")
            completion(false)
            return
        }

        let endpoint = "\(apiBaseURL)/reports"

        uploadRecord(endpoint: endpoint, method: operation.operationType, data: report) { result in
            switch result {
            case .success(let response):
                if let conflict = self.detectConflict(local: report, remote: response) {
                    self.handleConflict(conflict, operation: operation)
                    completion(false)
                } else {
                    self.markRecordSynced(type: "report", id: report.id)
                    self.storage.deleteSyncOperation(id: operation.id)
                    completion(true)
                }

            case .failure(let error):
                let retryCount = operation.retryCount + 1

                if retryCount >= self.maxRetries {
                    self.storage.updateSyncOperation(id: operation.id, retryCount: retryCount, error: error.localizedDescription)
                } else {
                    self.storage.updateSyncOperation(id: operation.id, retryCount: retryCount, error: error.localizedDescription)
                }

                completion(false)
            }
        }
    }

    private func syncPhoto(_ operation: SyncOperation, completion: @escaping (Bool) -> Void) {
        guard let photo = try? JSONDecoder().decode(PhotoRecord.self, from: operation.payload),
              let imageData = storage.getPhoto(id: photo.id) else {
            storage.updateSyncOperation(id: operation.id, retryCount: operation.retryCount + 1, error: "Failed to load photo")
            completion(false)
            return
        }

        let endpoint = "\(apiBaseURL)/photos/upload"

        uploadPhoto(endpoint: endpoint, photoId: photo.id, imageData: imageData) { result in
            switch result {
            case .success(let cloudUrl):
                // Update photo record with cloud URL
                self.updatePhotoCloudUrl(id: photo.id, cloudUrl: cloudUrl)
                self.markRecordSynced(type: "photo", id: photo.id)
                self.storage.deleteSyncOperation(id: operation.id)
                completion(true)

            case .failure(let error):
                let retryCount = operation.retryCount + 1

                if retryCount >= self.maxRetries {
                    self.storage.updateSyncOperation(id: operation.id, retryCount: retryCount, error: error.localizedDescription)
                } else {
                    self.storage.updateSyncOperation(id: operation.id, retryCount: retryCount, error: error.localizedDescription)
                }

                completion(false)
            }
        }
    }

    // MARK: - Network Operations

    private func uploadRecord<T: Encodable>(
        endpoint: String,
        method: String,
        data: T,
        completion: @escaping (Result<[String: Any], Error>) -> Void
    ) {
        guard let url = URL(string: endpoint) else {
            completion(.failure(NSError(domain: "Invalid URL", code: -1, userInfo: nil)))
            return
        }

        var request = URLRequest(url: url)
        request.httpMethod = method.uppercased()
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        // Add authentication token if available
        if let token = UserDefaults.standard.string(forKey: "auth_token") {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }

        do {
            request.httpBody = try JSONEncoder().encode(data)
        } catch {
            completion(.failure(error))
            return
        }

        URLSession.shared.dataTask(with: request) { data, response, error in
            if let error = error {
                completion(.failure(error))
                return
            }

            guard let httpResponse = response as? HTTPURLResponse else {
                completion(.failure(NSError(domain: "Invalid response", code: -1, userInfo: nil)))
                return
            }

            guard (200...299).contains(httpResponse.statusCode) else {
                let error = NSError(domain: "HTTP Error", code: httpResponse.statusCode, userInfo: nil)
                completion(.failure(error))
                return
            }

            guard let data = data,
                  let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
                completion(.failure(NSError(domain: "Invalid JSON response", code: -1, userInfo: nil)))
                return
            }

            completion(.success(json))
        }.resume()
    }

    private func uploadPhoto(
        endpoint: String,
        photoId: String,
        imageData: Data,
        completion: @escaping (Result<String, Error>) -> Void
    ) {
        guard let url = URL(string: endpoint) else {
            completion(.failure(NSError(domain: "Invalid URL", code: -1, userInfo: nil)))
            return
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"

        let boundary = UUID().uuidString
        request.setValue("multipart/form-data; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")

        if let token = UserDefaults.standard.string(forKey: "auth_token") {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }

        var body = Data()

        // Add photo ID
        body.append("--\(boundary)\r\n".data(using: .utf8)!)
        body.append("Content-Disposition: form-data; name=\"photo_id\"\r\n\r\n".data(using: .utf8)!)
        body.append("\(photoId)\r\n".data(using: .utf8)!)

        // Add image data
        body.append("--\(boundary)\r\n".data(using: .utf8)!)
        body.append("Content-Disposition: form-data; name=\"image\"; filename=\"photo.jpg\"\r\n".data(using: .utf8)!)
        body.append("Content-Type: image/jpeg\r\n\r\n".data(using: .utf8)!)
        body.append(imageData)
        body.append("\r\n".data(using: .utf8)!)
        body.append("--\(boundary)--\r\n".data(using: .utf8)!)

        request.httpBody = body

        URLSession.shared.dataTask(with: request) { data, response, error in
            if let error = error {
                completion(.failure(error))
                return
            }

            guard let httpResponse = response as? HTTPURLResponse,
                  (200...299).contains(httpResponse.statusCode) else {
                completion(.failure(NSError(domain: "Upload failed", code: -1, userInfo: nil)))
                return
            }

            guard let data = data,
                  let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let cloudUrl = json["url"] as? String else {
                completion(.failure(NSError(domain: "Invalid response", code: -1, userInfo: nil)))
                return
            }

            completion(.success(cloudUrl))
        }.resume()
    }

    // MARK: - Conflict Resolution

    private func detectConflict<T: Codable>(local: T, remote: [String: Any]) -> ConflictInfo? {
        // Parse remote lastModified timestamp
        guard let remoteModifiedTimestamp = remote["last_modified"] as? TimeInterval else {
            return nil
        }

        let remoteModified = Date(timeIntervalSince1970: remoteModifiedTimestamp)

        // Extract local lastModified using reflection (simplified for demo)
        let mirror = Mirror(reflecting: local)
        guard let localModified = mirror.children.first(where: { $0.label == "lastModified" })?.value as? Date else {
            return nil
        }

        // Conflict if remote was modified after local
        if remoteModified > localModified {
            return ConflictInfo(
                localTimestamp: localModified,
                remoteTimestamp: remoteModified,
                localData: local,
                remoteData: remote
            )
        }

        return nil
    }

    private func handleConflict(_ conflict: ConflictInfo, operation: SyncOperation) {
        // Default strategy: Last-Write-Wins (remote wins)
        print("Conflict detected - remote is newer, applying remote changes")

        // Store conflict for user review if needed
        let conflictRecord = ConflictRecord(
            id: UUID().uuidString,
            operationId: operation.id,
            recordType: operation.recordType,
            recordId: operation.recordId,
            localTimestamp: conflict.localTimestamp,
            remoteTimestamp: conflict.remoteTimestamp,
            status: "auto_resolved",
            resolution: "remote_wins",
            timestamp: Date()
        )

        saveConflictRecord(conflictRecord)

        // Update local with remote data
        applyRemoteChanges(operation: operation, remoteData: conflict.remoteData)
    }

    private func saveConflictRecord(_ record: ConflictRecord) {
        // Store conflict in database for audit trail
        storage.setSyncMetadata(key: "conflict_\(record.id)", value: record.id)
    }

    private func applyRemoteChanges(operation: SyncOperation, remoteData: [String: Any]) {
        // Apply remote changes to local database
        // Implementation depends on record type
        print("Applying remote changes for \(operation.recordType) \(operation.recordId)")

        // Mark operation as completed
        storage.deleteSyncOperation(id: operation.id)
    }

    // MARK: - Helper Methods

    private func markRecordSynced(type: String, id: String) {
        // Update record sync status in database
        storage.setSyncMetadata(key: "\(type)_\(id)_sync_status", value: "synced")
    }

    private func updatePhotoCloudUrl(id: String, cloudUrl: String) {
        storage.setSyncMetadata(key: "photo_\(id)_cloud_url", value: cloudUrl)
    }

    func cancelSync() {
        isSyncing = false
        print("Sync cancelled")
    }

    // MARK: - Public API

    func forceSyncNow() {
        guard isOnline else {
            print("Cannot force sync - offline")
            return
        }

        startSync()
    }

    func getSyncStatus() -> [String: Any] {
        let pendingOps = storage.getPendingSyncOperations()

        return [
            "is_syncing": isSyncing,
            "is_online": isOnline,
            "pending_operations": pendingOps.count,
            "last_sync": storage.getSyncMetadata(key: "last_sync_timestamp") ?? "never"
        ]
    }

    func clearSyncQueue() {
        // Clear all pending operations (use with caution)
        let operations = storage.getPendingSyncOperations(limit: 1000)
        operations.forEach { storage.deleteSyncOperation(id: $0.id) }
    }
}

// MARK: - Supporting Types

struct ConflictInfo {
    let localTimestamp: Date
    let remoteTimestamp: Date
    let localData: Any
    let remoteData: [String: Any]
}

struct ConflictRecord {
    let id: String
    let operationId: String
    let recordType: String
    let recordId: String
    let localTimestamp: Date
    let remoteTimestamp: Date
    let status: String
    let resolution: String
    let timestamp: Date
}

// MARK: - Notification Names

extension Notification.Name {
    static let networkStatusChanged = Notification.Name("networkStatusChanged")
    static let syncCompleted = Notification.Name("syncCompleted")
    static let syncFailed = Notification.Name("syncFailed")
}
